src/Fixer/Whitespace/StatementIndentationFixer.php

```diff
@@ -0,0 +1,600 @@
+<?php
+
+/*
+ * This file is part of PHP CS Fixer.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *     Dariusz Rumiński <dariusz.ruminski@gmail.com>
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace PhpCsFixer\Fixer\Whitespace;
+
+use PhpCsFixer\AbstractFixer;
+use PhpCsFixer\Fixer\WhitespacesAwareFixerInterface;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\Preg;
+use PhpCsFixer\Tokenizer\Analyzer\AlternativeSyntaxAnalyzer;
+use PhpCsFixer\Tokenizer\CT;
+use PhpCsFixer\Tokenizer\Token;
+use PhpCsFixer\Tokenizer\Tokens;
+
+final class StatementIndentationFixer extends AbstractFixer implements WhitespacesAwareFixerInterface
+{
+    private $bracesFixerCompatibility;
+
+    /**
+     * @param bool $bracesFixerCompatibility
+     */
+    public function __construct($bracesFixerCompatibility = false)
```

Comment:
Please remove constructor parameter from this and other fixers.
Fixers should become services and receive only dependencies via constructor.

This look like configuration (the option should state what the trigger changes in behaviour of fixer, not if it's compat mode with old fixer or not)

Reply:
I used these options to drop behaviors from `braces` that were too specific and opinionated IMO and that I didn't want to keep. The new fixers don't have those behaviors at all, unless executed via `braces` for backward compatibility.

Do you want to keep the `braces` specific behavior with the new fixers?

Reply:
what are the `specific behaviors` of `braces`?
can't understand it by looking at `bracesFixerCompatibility` name :(
I think we should decide one by one.

Same time, fixer is public, please no constructor parameter like this.
If we would decide that sth is too weird and we keep it as compatibility only and not as a proper configuration option, then create a method like `/** @internal */ enableBracesCompatibilityMode(): void {...}`

Reply:
Here is the list of `braces` specific behaviors, with test failures from `BracesFixerTest` that cover them.

<details><summary>1. In <code>control_structure_braces</code>: when a control structure has no braces and contains an alternative syntax control structure, do not add braces on the outer control structure</summary>

```diff
-<?php if ($a) foreach ($b as $c): ?> X <?php endforeach; ?>
+<?php if ($a) {
+    foreach ($b as $c): ?> X <?php endforeach;
+} ?>
```

Maybe we could improve this by using the alternative syntax on the outer control structure (but out of scope here I'd say):

```diff
-<?php if ($a) foreach ($b as $c): ?> X <?php endforeach; ?>
+<?php if ($a): foreach ($b as $c): ?> X <?php endforeach; endif; ?>
```

</details>

<details><summary>2. In <code>blank_lines_inside_block</code>: when a curly braces block starts with blank lines followed by a comment, don't remove the blank lines</summary>

```diff
 <?php
 if (true) {
-
     //  The blank line helps with legibility in nested control structures
     if (true) {
         // if body
     }
 
     // if body
 }
```

</details>

<details><summary>3. In <code>statement_indentation</code>: when a comment looks like commented-out code, don't indent it</summary>

```diff
 <?php
 function foo()
 {
     $a = 1;
-//    if ($a === 'bar') {
-//        return [];
-//    }
+    //    if ($a === 'bar') {
+    //        return [];
+    //    }
     // we will return sth
     return $a;
 }
```

</details>

<details><summary>4. In <code>statement_indentation</code>: when a single line comment is part of a multiline message, don't indent it</summary>

```diff
 <?php
 function foo()
 {
     $bar = 1;                   // multiline ...
-                                // ... comment
+    // ... comment
     $baz  = 2;                  // next comment
 }
```

</details>

<details><summary>5. In <code>statement_indentation</code>: assume code that is not inside any curly braces block is correctly indented</summary>

```diff
 <?php
-    if ($test) { //foo
-        echo 1;
-    }
+if ($test) { //foo
+    echo 1;
+}
```

</details>

Reply:
ref https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/4884#issuecomment-1122604242

I believe we should not change code to START using an alternative syntax
about 1&2 @julienfalque , I'm OK to keep it nice way in new fixers, as we merged them, and actually _any_ way in braces. let's get this deprecation merged asap, even if maybe code is not perfect - I will name it my mistake to aim to have it nice here (eg no config passed to constructor). We should aim to limit the impact of code-smell to other fixers even if it partially affect braces fixer in way we didn't wanted initially, and then we should bump major version to get rid of braces rule.

Reply:
> let's figure out how to minify remaining work to have this PR merged

I'm preparing a new PR to extract parts from this one.

> then let's open PR for v4 cleanup of legacy `Braces` fixer and it's spill-over (`$bracesFixerCompatibility`) to fresh rules

This work already exists but will need an epic rebase at some point: #4885

Reply:
`$bracesFixerCompatibility` dropped for `control_structure_braces` and `blank_lines_inside_block`.